trainer: Initial Documentation for Kubeflow Trainer V2

andreyvelich opened the pull request:
Fixes: https://github.com/kubeflow/training-operator/issues/2214

This is initial version for Kubeflow Training V2 docs.
Please let me know what do you think.

TODOs:
- [x] Add working Getting Started example.
- [x] Fix the installation scripts.
- [x] Add new logo of Kubeflow Training.
- [x] Rename Kubeflow Training Operator -> Kubeflow Training everywhere.

/cc @kubeflow/wg-training-leads @kubeflow/release-team @hbelmiro @varodrig @jbottum @varshaprasad96 @akshaychitneni @helenxie-bit @Electronic-Waste @saileshd1402 @seanlaii @deepanker13 @astefanutti @shravan-achar @kannon92 @droctothorpe @sandipanpanda @vsoch @franciscojavierarceo @Syulin7 @StefanoFioravanzo @kuizhiqing

andreyvelich commented on content/en/docs/components/training/legacy-v1/_index.md:
```diff
@@ -0,0 +1,11 @@
+weight = 999
++++
+
+{{% alert title="Old Version" color="warning" %}}
+This page is about **Kubeflow Training V1**, please see the [V2 documentation](/docs/components/training) for the latest information.
+
+Please follow [this guide for migrating to Kubeflow Training V2](/docs/components/training/admin-guides/migration)
+{{% /alert %}}
```
Let me know if that message looks good to you @kubeflow/wg-training-leads @rimolive @varodrig @hbelmiro @StefanoFioravanzo.
If yes, I will add it to all Kubeflow Training V1 docs, similar to KFP: https://www.kubeflow.org/docs/components/pipelines/legacy-v1/overview/quickstart/

Electronic-Waste commented on content/en/docs/components/training/admin-guides/installation.md:
```diff
@@ -0,0 +1,58 @@
+
+These are the minimal requirements to install the Kubeflow Training control plane:
+
+- Kubernetes >= 1.27
+- `kubectl` >= 1.27
```
AFAIK, we've removed our support for 1.27: https://github.com/kubeflow/training-operator/blob/1dfa40c12516fc9eb2ce12c5ef52da7d46670457/.github/workflows/unittests.yaml#L21

andreyvelich replied:
Good point, let me update it.

Electronic-Waste commented on content/en/docs/components/training/admin-guides/installation.md:
````diff
@@ -0,0 +1,58 @@
+
+```
+
+## Installing the Kubeflow Training Runtimes
````
I think, it will be better if we could provide a standalone installation guide:)

andreyvelich replied:
Actually, I am planning to refactor our manifests given that the Cluster Training Runtime needs to be installed after the manager.
I will soon submit a PR.

astefanutti replied:
I would guess installing the cluster training runtime requires the CRDs to be installed first, more than the manager to be deployed. Projects tend to separate the steps to install CRDs first, then the rest of the manifests.

andreyvelich replied:
Actually, we also need to install manager before we deploy the CTR, since we perform validation and mutation via webhook.

Electronic-Waste commented on content/en/docs/components/training/legacy-v1/installation.md:
```diff
@@ -0,0 +1,99 @@
+
+These are the minimal requirements to install the Training Operator:
+
+- Kubernetes >= 1.27
+- `kubectl` >= 1.27
+- Python >= 3.7
```
Same as above

Electronic-Waste reviewed (commented):
@andreyvelich  Great contributions! I left the initial comments for you.

vsoch commented on content/en/docs/components/training/admin-guides/installation.md:
````diff
@@ -0,0 +1,58 @@
+If you don't have Kubernetes cluster, you can quickly create one locally using [Kind](https://kind.sigs.k8s.io/docs/user/quick-start#installing-with-a-package-manager):
+
+```bash
+brew install kind
````
You assume users have Mac?

andreyvelich replied:
It is a good point, I just assume that many Data Scientists who use Mac might be using `brew` already.
Any other thoughts here @vsoch @kubeflow/wg-training-leads @astefanutti ?

astefanutti replied:
I agree we can assume Mac users might be using Brew, but at that stage I think it's about documenting how to create a Kind cluster, not installing Kind, which is provided by the link. So here keeping `kind create cluster` may be just enough and not exclude non-Mac users :)

vsoch commented on content/en/docs/components/training/admin-guides/installation.md:
````diff
@@ -0,0 +1,58 @@
+TODO (andreyvelich): Change the link once V1 is removed.
+
+```bash
+kubectl apply --server-side -k "github.com/kubeflow/training-operator.git/manifests/v2/overlays/manager?ref=master"
````
Don't most copy paste a full URL that starts with https?

andreyvelich replied:
Actually, this is how you can use remote URL with kubectl and kustomize (e.g. they accept the SSH url: github.com:kubeflow/training-operator.git )

vsoch commented on content/en/docs/components/training/installation.md:
```diff
@@ -1,99 +1,41 @@
 ## Prerequisites
 
-These are the minimal requirements to install the Training Operator:
+Ensure that you have access to a Kubernetes clusters with the Kubeflow Training
```
```suggestion
Ensure that you have access to a Kubernetes cluster with the Kubeflow Training
```

vsoch commented on content/en/docs/components/training/installation.md:
```diff
@@ -1,99 +1,41 @@
-These are the minimal requirements to install the Training Operator:
+Ensure that you have access to a Kubernetes clusters with the Kubeflow Training
+control plane installed. If it is not set up yet, follow
+[the installation guide](/docs/components/training/admin-guides/installation) to quickly deploy
+Kubeflow Training on a local Kind cluster.
```
```suggestion
[the installation guide](/docs/components/training/admin-guides/installation) to deploy
Kubeflow Training.
```
No reason it needs to be kind, and if there are webhooks it won't be that quick :P

andreyvelich replied:
I intentionally add this message to tell that it is super easy to deploy Kubernetes locally, and quickly try Kubeflow Training. I don't want to scare our ML users with "Kubernetes" dependency.
@kubeflow/wg-training-leads @vsoch @astefanutti @franciscojavierarceo @StefanoFioravanzo Any thoughts on how we can phrase it better in docs ?

astefanutti replied:
I agree with @vsoch suggestion, it doesn't seem there is a need to be too specific here, "quickly deploy Kubeflow Trainer" straight makes it even less scary here and do not exclude non-Kind options.

andreyvelich replied:
@vsoch @astefanutti ?

vsoch commented on content/en/docs/components/training/installation.md:
```diff
@@ -1,99 +1,41 @@
+## Installing the Kubeflow SDK
 
-## Installing the Training Operator
+You can chose between installing the latest stable release of the development version from
```
```suggestion
You can chose between installing the latest stable release or the development version from
```

vsoch commented on content/en/docs/components/training/legacy-v1/getting-started.md:
```diff
@@ -0,0 +1,158 @@
+
+You need to install the following components to run examples:
+
+- The Training Operator control plane [installed](/docs/components/training/legacy-v1/installation/#installing-the-control-plane).
```
I would move the link here to be around the name of the components and nix "installed" - it doesn't match the previous sentence that leads into it. "You need to install" ... "installed"

andreyvelich replied:
Actually, we don't want to modify the Training Operator V1 doc as part of this PR, since it focuses on Kubeflow Training V2 doc. Does it look good @vsoch ?

vsoch commented on content/en/docs/components/training/legacy-v1/getting-started.md:
```diff
@@ -0,0 +1,158 @@
+You can create your first Training Operator distributed PyTorchJob using the Python SDK. Define the
+training function that implements end-to-end model training. Each Worker will execute this
+function on the appropriate Kubernetes Pod. Usually, this function contains logic to
+download dataset, create model, and train the model.
```
```suggestion
download a dataset, create a model, and train the model.
```

vsoch commented on content/en/docs/components/training/legacy-v1/installation.md:
```diff
@@ -0,0 +1,99 @@
+
+## Installing the Training Operator
+
+You need to install the Training Operator control plane and Python SDK to create training jobs.
```
You call it "control plane" several times, but I think you mean controller? The controller might be installed to the control plane but is it a control plane as we understand them for Kubernetes?

andreyvelich replied:
The control plane here means similar to Kubernetes control plane: https://kubernetes.io/docs/concepts/overview/components/#control-plane-components
For now, Training Operator control plane contains controller and webhook.

vsoch commented on content/en/docs/components/training/legacy-v1/installation.md:
```diff
@@ -0,0 +1,99 @@
+### Installing the Python SDK
+
+The Training Operator [implements a Python SDK](https://pypi.org/project/kubeflow-training/)
+to simplify creation of distributed training and fine-tuning jobs for Data Scientists.
```
```suggestion
to simplify creation of distributed training and fine-tuning jobs.
```
Why just data scientists?

vsoch commented on content/en/docs/components/training/legacy-v1/installation.md:
````diff
@@ -0,0 +1,99 @@
+Otherwise, you can also install the Training SDK using the specific GitHub commit, for example:
+
+```shell
+pip install git+https://github.com/kubeflow/training-operator.git@7345e33b333ba5084127efe027774dd7bed8f6e6#subdirectory=sdk/python
````
What about a release version that is a little more pinned than a branch, but not as fine grained as needing a specific  commit?

andreyvelich replied:
They can select it from PyPi directly:
```
pip install kubeflow-training==1.9.0rc0
```

vsoch commented on content/en/docs/components/training/legacy-v1/overview.md:
```diff
@@ -0,0 +1,79 @@
+
+- **The Training Operator is extensible and portable.**
+
+You can deploy Training Operator on any cloud where you have Kubernetes cluster and you can
```
```suggestion
You can deploy the Training Operator on any cloud where you have Kubernetes cluster and you can
```

vsoch commented on content/en/docs/components/training/overview.md:
```diff
@@ -1,79 +1,89 @@
 +++
 title = "Overview"
-description = "An overview of the Training Operator"
+description = "An overview of the Kubeflow Training"
```
```suggestion
description = "An overview of Kubeflow Training"
```
or THE Kubeflow Training... API?

andreyvelich replied:
Kubeflow Training is the official product/project name.
@vsoch @franciscojavierarceo @rimolive @varodrig @hbelmiro @kubeflow/wg-training-leads @thesuperzapper  Should we use the definite article everywhere or not ?

vsoch commented on content/en/docs/components/training/overview.md:
```diff
@@ -1,79 +1,89 @@
 
-{{% stable-status %}}
+{{% alert title="Note" color="dark" %}}
+The Kubeflow Training V2 is currently in <strong>alpha</strong> status, and APIs may change.
```
This is missing what Kubeflow Training is - an API? Something else? It sounds strange to say "The Kubeflow Training."

andreyvelich replied:
Let me say Kubeflow Training project, WDYT @vsoch ?

vsoch commented on content/en/docs/components/training/overview.md:
```diff
@@ -1,79 +1,89 @@
-PyTorch, TensorFlow, XGBoost, JAX, and others.
+## What is the Kubeflow Training
+
+The Kubeflow Training is a Kubernetes-native project for large language models (LLMs) fine-tuning
```
Same here - need a descriptor with "The"

vsoch replied:
```suggestion
The Kubeflow Training project is a Kubernetes-native project for large language models (LLMs) fine-tuning
```

vsoch commented on content/en/docs/components/training/overview.md:
```diff
@@ -1,79 +1,89 @@
-with the Training Operator to orchestrate their ML training on Kubernetes.
+with the Kubeflow Training to orchestrate their ML training on Kubernetes.
+
+The Kubeflow Training allows you effortlessly develop your LLMs with the Kubeflow Python SDK and
```
Ditto (and above this)

vsoch commented on content/en/docs/components/training/overview.md:
```diff
@@ -1,79 +1,89 @@
-supports running Message Passing Interface (MPI) on Kubernetes which is heavily used for HPC.
-The Training Operator implements the V1 API version of MPI Operator. For the MPI Operator V2 version,
-please follow [this guide](/docs/components/training/user-guides/mpi/) to install MPI Operator V2.
+The Kubeflow Training is designed for two primary user personas, each with specific resources and
```
Ditto

vsoch commented on content/en/docs/components/training/overview.md:
```diff
@@ -1,79 +1,89 @@
-various distributed training strategies for different ML frameworks.
+### User Personas
+
+The Kubeflow Training documentation is separated between these user personas:
```
Ditto

vsoch commented on content/en/docs/components/training/overview.md:
```diff
@@ -1,79 +1,89 @@
+
+## Why use the Kubeflow Training
+
+The Kubeflow Training supports key phases on the AI/ML lifecycle, including model training and LLMs
```
Ditto

andreyvelich commented on content/en/_redirects:
```diff
@@ -337,4 +337,19 @@ docs/started/requirements/                    /docs/started/getting-started/
+/docs/components/training/user-guides/paddle                 /docs/components/training/legacy-v1/user-guides/paddle
+/docs/components/training/user-guides/prometheus             /docs/components/training/legacy-v1/user-guides/prometheus
+/docs/components/training/user-guides/tensorflow             /docs/components/training/legacy-v1/user-guides/tensorflow
+/docs/components/training/user-guides/xgboost                /docs/components/training/legacy-v1/user-guides/xgboost
```
We discussed with @thesuperzapper how to structure the new docs, and he suggested that we put new user guides under
```/docs/components/training/user-guides-v2```
For example: 

That will allow us to redirect the V1 docs (e.g. TFJob) to the correct location.

What do you think about it @kubeflow/wg-training-leads ?

thesuperzapper replied:
@andreyvelich make sure you also redirect the "index" pages e.g.

- `/docs/components/training/explanation/` -> `/docs/components/training/legacy-v1/explanation/`
- `/docs/components/training/user-guides/` -> `/docs/components/training/legacy-v1/user-guides/`
- `/docs/components/training/reference/` -> `/docs/components/training/legacy-v1/reference/`

Also, I am not sure if it matters, but all other redirects use a trailing `/` slash.

andreyvelich replied:
Good catch, let me fix that.

franciscojavierarceo commented on content/en/docs/components/training/_index.md:
```diff
@@ -1,5 +1,5 @@
-title = "Training Operator"
-description = "Documentation for Kubeflow Training Operator"
-weight = 70
+title = "Kubeflow Training"
```
What do you think about calling it "Kubeflow Trainer"? 🤔

andreyvelich replied:
Hmm, this could be one of the options to name this project Kubeflow Trainer/KFTrainer.
@franciscojavierarceo Please can you explain why do you prefer this project name over the KFTraining ?

franciscojavierarceo replied:
It would follow HuggingFace's and Lightning.ai's, "trainer" convention: https://huggingface.co/docs/transformers/en/main_classes/trainer
https://lightning.ai/docs/pytorch/stable/common/trainer.html

andreyvelich replied:
cc @astefanutti @kubeflow/wg-training-leads @Electronic-Waste @deepanker13 @saileshd1402 @seanlaii

astefanutti replied:
So at the moment, from an end-user standpoint, the story goes like this:

> In Kubeflow training, I have a `TrainingClient` I can use to `train` (a model on Kubernetes) by providing a `Trainer` config.

If I understand it correctly, that would change to something like:

> In Kubeflow Trainer, I have a `TrainerClient` I can use to `train` (a model on Kubernetes) by providing a `Trainer` config.

Following the analogy with HuggingFace or Lightning, `Trainer` is class in the Transformers (resp. Lightning) project. We could also take the `SFTTrainer` class from the TRL project as another example. For the first one, that gives:

> In HF Tranformers, I have a `Trainer` I can use to `train` (a model) by providing some parameters.

I find it rare when projects are named based on API constructs, and instead use a more functional (or creative) name, that's less narrow than a class name, so it's more representative of the scope or problem it addresses, and also more durable / resistant to change.

If the above is correct,  I'd be inclined to keep the project name focused on the WHY/WHAT, more than on the HOW. But I could see why we could want to change the API names.

Naming is hard :)

andreyvelich replied:
>In Kubeflow Trainer, I have a TrainerClient I can use to train (a model on Kubernetes) by providing a Trainer config.

Yes, that is correct.

I think, we should find a project name that gives users a clear idea why and when they should use it, right ?
For example:
- Kubeflow Trainer: to fine-tune of train ML models.
- Kubeflow Tuner: to optimize model hyperparameters.
- Kubeflow Model Registry: to create model artifactory
- Kubeflow Pipelines: to build pipeline using the above products/services.

If we keep the project name as: Kubeflow Training and Kubeflow Tuning, does it make clear for users when they should use it ?

astefanutti replied:
> I think, we should find a project name that gives users a clear idea why and when they should use it, right ?
For example:
>
> * Kubeflow Trainer: to fine-tune of train ML models.
> * Kubeflow Tuner: to optimize model hyperparameters.
> * Kubeflow Model Registry: to create model artifactory
> * Kubeflow Pipelines: to build pipeline using the above products/services.

Right, that looks pretty good. There would clearly be value in applying that approach across the board.
Train, Tune, Serve, Registry, Pipelines, and maybe Eval, Store could look like a good line-up.

> If we keep the project name as: Kubeflow Training and Kubeflow Tuning, does it make clear for users when they should use it ?

I guess we would need to ask users directly to get an answer to that question :)

franciscojavierarceo replied:
+1 looks great

chasecadet replied:
@andreyvelich @franciscojavierarceo   I think Kubeflow trainer sounds like poor English. It's not training Kubeflow. It's a trainer on Kubeflow and really serves the upstream frameworks. Operator was nice due to the fact we were indeed spinning up resources and managing the lifecycles of frameworks. We created PyTorch clusters, ran a job, and could upgrade it etc.. all "operational" expertise we offloaded.  Unless this is a trainer python package directly training something like a HuggingFace, I would shy away from trainer. This could also confuse the community more around what we are and overload more terms (something we in tech love to do). My suggestions 

* Kubeflow Training Manager
* Kubeflow Training System
* Kubeflow Training Module
* Kubeflow Training Runtime
* Kubeflow Training Service 


Thoughts? Am I crazy?

vsoch replied:
It also makes me think of Trainers, as in sneakers.

andreyvelich replied:
I think, the goal of naming it Kubeflow Trainer is to make it clear for data scientists and ML Engineers that they can use this project for **AI model pre-training and post-training.**

The cluster admins can always check the operator guides to understand the details.

>It's a trainer on Kubeflow and really serves the upstream frameworks. Operator was nice due to the fact we were indeed spinning up resources and managing the lifecycles of frameworks. 

I think, ML Users don't need to know about it. They can just use it to scale their ML code across many GPUs.

>Unless this is a trainer python package directly training something like a HuggingFace, I would shy away from trainer.

This will be actually the trainer python package that ML users will use as I described above: https://github.com/kubeflow/website/pull/3958#discussion_r1917101576
ML users will interact with Kubeflow Trainer APIs using the `kubeflow` python SDK.

Which means we are going to have the consistent story for the entire ML lifecycle:
```py
from kubeflow.trainer import TrainerClient
from kubeflow.optimizer import OptimizerClient <--- This is Katib project 
from kubeflow.evaluator import EvaluatorClient <--- This could be a new project under Kubeflow

TrainerClient().train(...)
OptimizerClient.optimize(...)
EvaluatorClient().evaluate(...)

```

And on top of all of these projects, we have KFP which allows ML users to orchestrate it:

```py
from kfp import dsl
import kfp

@dsl.component
def train():
    TrainerClient().train(...)

@dsl.component
def optimize():
     OptimizerClient.optimize(...)

@dsl.component
def eval():
    EvaluatorClient().evaluate(...)

@dsl.pipeline
def e2e_ml_pipeline():
    train()
    optimize()
    eval()
```

Does it make sense @chasecadet @vsoch ?
cc @johnugeorge @tenzen-y @franciscojavierarceo

chasecadet replied:
@andreyvelich Thanks for the response and showing me the math! I like that perspective. 
This is probably better done in person since we lose a lot doing async responses and I feel it actually takes longer for discussions beyond "LGTM" and our subject to attention spans and character limits.. but I digress. 

So let's set the scene here.
**The goal:** Move users away from thinking of Kubeflow as just a means to orchestrate pure infrastructure. We want to show a user we can go end to end with model training so we propose moving the Training Operator (A Kubeflow infra term) to a more MLE DS term *trainer* . We also seem to want to abstract away our infrastructure super powers. I'd argue that The infra orchestration is why someone comes to us (Its not like we are building a PyTorch competitor or Ray. We are orchestrating Kubernetes components) 

**The initial concern** is that trainer feels like we are training Kubeflow not training a model. We also are again, just orchestrating frameworks and compute.   Now if we called this the *Kubeflow Training Service*, *Kubeflow Optimization Service*, and *Kubeflow evaluation service* it feels like Kubeflow is doing work on our behalf beyond JUST training a model (we are more than just a single kernel etc..). We submit jobs to the service like you'd submit a work request to a company to perform a service. Also, Kserve serving a model endpoint in the cloud is "model as a service". We are leaning into the platform/cloud jargon and can start to build a story around submitting jobs/tasks to the Kubeflow platform services. Thoughts? Am I crazy?

andreyvelich replied:
>Its not like we are building a PyTorch competitor or Ray. We are orchestrating Kubernetes components)

I would argue to that, since Kubeflow component such as Katib or Training Operator is alternative to Ray Tune and Ray Train libraries.
E.g. it helps you to distribute AI workloads using Kubernetes native resources and ML framework native tools, such as `torchrun`, `mpirun`, `horovod`.

> is that trainer feels like we are training Kubeflow not training a model.

What do you mean by that, please can you explain if we say that: 
**Kubeflow is an Ecosystem of Open Source Projects**? 

Kubeflow Trainer, it's like Kubeflow Pipelines or Kubeflow Serving (e.g. KFServing)
.It is a product name. E.g. you can call it **KFTrainer**

>Now if we called this the Kubeflow Training Service

This is a good idea, but it would be hard to be consistent across all Kubeflow component.
For example, probably we don't want to rename Kubeflow Pipelines to Kubeflow Pipelines Service.

Also, enterprises might think that the Kubeflow Training Service will give them end-to-end AI/ML experience.
However, we have very narrow scope for this project. We will only do AI model training as part of this product.

@chasecadet Why you don't like the idea that Kubeflow Platform combines all of this products together to build production-ready AI/ML end-to-end Platform ?
E.g.

```
------------- Kubeflow Platform -------------
- Kubeflow Workspaces          <---- AI Model Development
- Kubeflow Spark               <---- AI Data Processing
- Kubeflow Trainer             <--- AI Model Training
- Kubeflow Optimizer           <--- AI Model Optimization
- Kubeflow Model Registry      <---- AI Model Management
- Kubeflow Pipelines           <--- Run ML pipelines using the above services
------------- Kubernetes -------------
```

I am happy to hear feedback from other community members on this.
cc @kubeflow/kubeflow-steering-committee  @kubeflow/wg-training-leads @kubeflow/wg-pipeline-leads @kubeflow/wg-notebooks-leads @kubeflow/wg-manifests-leads @kubeflow/release-team @kubeflow/wg-data-leads @kubeflow/wg-deployment-leads

chasecadet replied:
@andreyvelich  I don't like that Kubeflow platform combines the products together? 
Also,
> For example, probably we don't want to rename Kubeflow Pipelines to Kubeflow Pipelines Service.
Why not? What are you optimizing for and based on what demand? Have we opened this up to the greater community? Who are our "customers" so to speak and how would they want APIs/Components labeled? This seems like a big decision to make in a vacuum and thought leadership is service (Kelsey has for sure played this game well). What if we opened it up to the greater community? Posted some options on poles via socials we as a community publish? We can then be more data driven. I bet @StefanoFioravanzo has some opinions!

terrytangyuan replied:
One concern to me is that training operator is not limited to training use case but can be expanded to parallel computing (via MPIJob). Similarly, `PyTorchJob` is not limited to PyTorch trainer but more like [PyTorch distributed](https://pytorch.org/tutorials/beginner/dist_overview.html) which offers primitives to and abstractions for parallelism, sharding, and communications.

andreyvelich replied:
> @andreyvelich I don't like that Kubeflow platform combines the products together?

My points don't imply that Kubeflow products cannot function as standalone applications. However, for users interested in seeing how these individual open-source projects integrate seamlessly, the Kubeflow Platform provides a comprehensive, end-to-end machine learning experience."

>Why not? What are you optimizing for and based on what demand? Have we opened this up to the greater community?

I believe it will take our community significantly more time to discuss this thoroughly (approximately 1–2 years given the user base of Kubeflow Pipelines).
In the meantime, we should focus on releasing our current project.

>One concern to me is that training operator is not limited to training use case but can be expanded to parallel computing (via MPIJob). Similarly, PyTorchJob is not limited to PyTorch trainer but more like [PyTorch distributed](https://pytorch.org/tutorials/beginner/dist_overview.html) which offers primitives to and abstractions for parallelism, sharding, and communications.

That is correct, but right now it is out of scope of supported CRDs (TrainJob and TrainingRuntime). Theoretically, users can leverage TrainingRuntimes and TrainJob for distributed inference with MPI, but it would be better if we create dedicated CRDs for it. Also, our `kubeflow` Python SDK doesn't support it.

---

All in all, naming this project `Kubeflow Trainer` will help avoid user confusion, considering the following user journey:
1. **Cluster Operators** install Kubeflow Trainer controller manager into Kubernetes cluster.
2. **Cluster Operators** configure the required Training Runtimes for ML users.
3. **ML Users** use Kubeflow Python SDK to create TrainJob objects and interact with the Kubeflow Trainer APIs:
```python
from kubeflow.trainer import TrainerClient

# Get available runtimes.
TrainerClient().list_runtimes()

# Train my ML model
TrainerClient().train(
  func=train_func,
  func_args={"lr": 0.01}
  num_nodes=100
  resources_per_node={"gpu": 5}
)
```

We can always revisit the project name in the future if users tell us that this experience is bad.
What do you think ?

Electronic-Waste replied:
```
------------- Kubeflow Platform -------------
- Kubeflow Workspaces          <---- AI Model Development
- Kubeflow Spark               <---- AI Data Processing
- Kubeflow Trainer             <--- AI Model Training
- Kubeflow Optimizer           <--- AI Model Optimization
- Kubeflow Model Registry      <---- AI Model Management
- Kubeflow Pipelines           <--- Run ML pipelines using the above tools
------------- Kubernetes --------------------
```

I would support this naming convention, it would be more clear to users. than `training-operator` and `katib`.

tenzen-y replied:
I'm ok with `Kubeflow Trainer`. It's challenging to find the name that can cover the whole of the use cases (ML Training / HPC Computing) and all use personals (ClusterOperators / ML Engineers / Researchers / Backend Engineers). I believe that `Trainer` can flexibly cover all.

But I agree with here discussions since `Trainer` does not exactly cover the whole of things. Everyone talks about the project name based on different points of view (personas and workload specifications)

Note that Previously, we were seeking the Kubeflow `Batch` (or `Job`) as an alternative project name. 
However, we declined to introduce the `Batch` as a name since we believe the `Trainer` can imply MachineLearning semantics rather than `Batch`.

pdarshane commented on content/en/docs/components/katib/user-guides/hp-tuning/configure-experiment.md:
```diff
@@ -122,7 +122,7 @@ trialSpec:
 
 If you use `PyTorchJob` or other Training Operator jobs in your Trial template check
-[here](/docs/components/training/user-guides/tensorflow/#what-is-tfjob) how to set the annotation.
+[here](/docs/components/training/legacy-v1/user-guides/tensorflow/#what-is-tfjob) how to set the annotation.
```
Should there be a comma between "template" and "check"?

pdarshane commented on content/en/docs/components/training/contributor-guides/contributing.md:
```diff
@@ -0,0 +1,7 @@
+weight = 10
++++
+
+This document describes how to contribute to Kubeflow Training project.
```
I see inconsistent use of "This document" and "This page" throughout this PR. We can just start the sentence with something like "To contribute to the Kubeflow Training project... " or something similar. Just a suggestion.

andreyvelich replied:
I think, historically we've been using this in the beginning of the page:
```
This document describes how to ....
```
However, some pages have various messages.
@pdarshane From your point of view, how should we start our guides ?
cc @StefanoFioravanzo @varodrig

varodrig replied:
@andreyvelich  based on previous conversation, we will not be having a specific page for the contribution/community.

andreyvelich replied:
What do you mean @varodrig ? Even right now, individual Kubeflow projects have their own contributor guides:
- Kubeflow Training Operator: https://github.com/kubeflow/training-operator/blob/master/CONTRIBUTING.md
- Kubeflow Katib: https://github.com/kubeflow/katib/blob/master/CONTRIBUTING.md
- Spark Operator: https://www.kubeflow.org/docs/components/spark-operator/developer-guide/
I thought, we've just discussed whether we should cross-link these guides from the Kubeflow website.

varodrig replied:
I created an issue to reflect the conversation we had and made a few updates, feel free to make any suggestions
https://github.com/kubeflow/website/issues/3971
but the main idea is to not have individual pages on each project on the website, but continue one centralized place on the website and links to the git repos.

pdarshane commented on content/en/docs/components/training/getting-started.md:
````diff
@@ -1,158 +1,34 @@
-
-```python
-from kubeflow.training import TrainingClient
+Alternatively, you can install the latest Kubeflow Python SDK version directly
````
We can eliminate "you can" to reduce wordiness.

pdarshane commented on content/en/docs/components/training/legacy-v1/_index.md:
```diff
@@ -0,0 +1,11 @@
++++
+
+{{% alert title="Old Version" color="warning" %}}
+This page is about **Kubeflow Training V1**, please see the [V2 documentation](/docs/components/training) for the latest information.
```
Style guides typically recommend against using "please".

varodrig reviewed (commented):
Additionally, we need to:

- Update Katib documentation to point to v2
- Update Contributions page https://www.kubeflow.org/docs/about/contributing/
- Update Community Page https://www.kubeflow.org/docs/about/community/
@andreyvelich

andreyvelich commented on content/en/docs/components/trainer/getting-started.md:
```diff
@@ -0,0 +1,34 @@
+
+## Getting Started with PyTorch
+
+TODO (andreyvelich): Add example from the Notebook
```
I will add the getting started example once we finish this PR with @astefanutti: https://github.com/kubeflow/training-operator/pull/2387

andreyvelich commented on content/en/docs/components/trainer/legacy-v1/_index.md:
```diff
@@ -0,0 +1,12 @@
+This page is about **Kubeflow Training Operator V1**, for the latest information check
+[the Kubeflow Trainer V2 documentation](/docs/components/trainer).
+
+Follow [this guide for migrating to Kubeflow Trainer V2](/docs/components/trainer/operator-guides/migration)
```
Maybe we could just say:
Follow this guide for migrating to the new Kubeflow Trainer project.
WDYT @varodrig @kubeflow/wg-training-leads ?

andreyvelich commented on content/en/docs/components/trainer/legacy-v1/installation.md:
```diff
@@ -96,4 +96,4 @@ pip install -U "kubeflow-training[huggingface]"
 ## Next steps
 
-Run your first Training Operator Job by following the [Getting Started guide](/docs/components/training/getting-started/).
+Run your first Training Operator Job by following the [Getting Started guide](/docs/components/trainer/getting-started/).
```
Good catch!

andreyvelich commented on content/en/docs/components/trainer/legacy-v1/reference/fine-tuning.md:
```diff
@@ -5,13 +5,13 @@ weight = 10
 In the following diagram you can see how `train` Python API works:
 
-<img src="/docs/components/training/images/fine-tune-llm-api.drawio.svg"
+<img src="/docs/components/trainer/legacy-v1/images/fine-tune-llm-api.drawio.svg"
```
Do we really need to do it here ?
I think, Access Mode is stable feature in Kubernetes since v1.18, so we won't expect any changes to this section.
Since we can support newer version of Kubernetes in the future releases of Training Operator (e.g. v1.9.2), it would be hard to always update these links.

andreyvelich commented on content/en/docs/components/trainer/operator-guides/installation.md:
````diff
@@ -0,0 +1,59 @@
+kubectl apply --server-side -k "https://github.com/kubeflow/training-operator.git/manifests/v2/overlays/manager?ref=master"
+```
+
+TODO (andreyvelich): Change the name.
````
Let me remove it for now, I will update it in the followup PRs once we merge this: https://github.com/kubeflow/training-operator/pull/2389

andreyvelich commented on content/en/docs/components/trainer/operator-guides/installation.md:
```diff
@@ -0,0 +1,59 @@
+
+TODO (andreyvelich): Change the name.
+
+Ensure that the JobSet and Trainer controller manager pods are running:
```
Yes, I will update it in the followup PRs.

andreyvelich commented on content/en/docs/components/trainer/operator-guides/migration.md:
```diff
@@ -0,0 +1,7 @@
++++
```
Good idea, I added it.

andreyvelich commented on content/en/docs/started/architecture.md:
@varodrig Where do you want to update it ?

varodrig replied:
@andreyvelich it's on the https://www.kubeflow.org/docs/started/architecture/#kubeflow-apis-and-sdks
let me know if this helps

andreyvelich commented on content/en/docs/components/trainer/overview.md:
```diff
@@ -0,0 +1,90 @@
+Kubeflow Trainer is designed for two primary user personas, each with specific resources and
+responsibilities:
+
+<img src="/docs/components/trainer/images/user-personas.drawio.svg"
```
@varodrig Please can you point to the diagram which is broken ?

varodrig replied:
I just noticed the image is fine but when showing on the web page is not. So, one of the main problems is that the background is black, when loading on the page, because the background is white all the white content from personas titles, arrows, logos titles are not showing. the kubeflow python SDK is the one that is not showing the whole image properly.
![Screenshot 2025-01-20 at 6 55 28 pm](https://github.com/user-attachments/assets/2cb8a00f-278b-4baa-a67a-0bfa0cf44e70)
.

andreyvelich replied:
Which browser do you use @varodrig ?
@thesuperzapper Do you know what is the right way to export images from drawio to avoid such problems ?

varodrig replied:
I'm using Chrome

tenzen-y replied:
I have same problems with @varodrig and I am using chrome as well

andreyvelich replied:
@sftim @Arhell Do you know why are we getting this issue with Hugo ?
Does Kubernetes also use drawio for the diagrams ?

sftim replied:
Kubernetes docs mostly don't use draw.io (some do, there are lots of diagrams). See https://kubernetes.io/docs/contribute/style/diagram-guide/ for what Kubernetes recommends.

varodrig replied:
I still have the same issues but only with these two images. the rest are ok

varodrig replied:
now works fine. thanks for fixing it. @andreyvelich

andreyvelich commented on content/en/docs/components/trainer/legacy-v1/installation.md:
Actually, the latest version of Kubeflow Platform 1.10 will also include Training Operator v1.

saileshd1402 commented on content/en/docs/components/trainer/getting-started.md:
````diff
@@ -0,0 +1,34 @@
+Install the Kubeflow Python SDK to interact with Kubeflow Trainer APIs:
+
+```bash
+pip install kubeflow
````
Currently when running this command, I see that the V1 SDK is installed. Should we wait for it to be updated to add it here, or is it fine keep it?

andreyvelich replied:
Yes, good point!
Let me remove it until we release SDK to PyPI and keep only command that installs SDK from the source.

saileshd1402 commented on content/en/docs/components/trainer/legacy-v1/user-guides/paddle.md:
```diff
@@ -16,7 +16,7 @@ Istio [automatic sidecar injection](https://istio.io/v1.3/docs/setup/additional-
 In order to get it running, it needs annotation `sidecar.istio.io/inject: "false"`
```
Do we need to be worried about links to examples/repo in documentation since we are moving V1 to a separate branch? For example: [line 23](https://github.com/kubeflow/website/pull/3958/files#diff-a87926ee99b436d97d499d1475653d8d46155ca532c3c051ef88c1c3eaaef648R23), [line 29](https://github.com/kubeflow/website/pull/3958/files#diff-a87926ee99b436d97d499d1475653d8d46155ca532c3c051ef88c1c3eaaef648R29) of this file. This would apply to most example job docs AFAIK

andreyvelich replied:
Oh, it is a great find! I think, we should link the examples from the `release-1.9` branch.
Let me update it.

saileshd1402 reviewed (commented):
LGTM! Just had a few minor doubts

varodrig commented on content/en/docs/components/trainer/contributor-guides/community.md:
```diff
@@ -0,0 +1,5 @@
++++
+title = "Community Guide"
```
I created an issue to reflect the conversation we had and made a few updates, feel free to make any suggestions
https://github.com/kubeflow/website/issues/3971
but the main idea is to not have individual pages on each project on the website, but continue one centralized place on the website and links to the git repos.

astefanutti commented on content/en/docs/components/trainer/getting-started.md:
```diff
@@ -0,0 +1,29 @@
+Ensure that you have access to a Kubernetes cluster with Kubeflow Trainer
+control plane installed. If it is not set up yet, followÍ
+[the installation guide](/docs/components/trainer/operator-guides/installation) to quickly deploy
+Kubeflow Trainer on your local Kind cluster.
```
```suggestion
Kubeflow Trainer.
```
It may be just better straight. The provided link gives the specifics.

Electronic-Waste reviewed (commented):
Just a few comments:)

andreyvelich commented on content/en/docs/started/architecture.md:
```diff
@@ -138,8 +138,8 @@ See the following sets of reference documentation:
   language (DSL).
-- [Training Operator Python SDK](https://github.com/kubeflow/training-operator/blob/86e0df17db715543b366e885c9ae659aa1342c8e/sdk/python/kubeflow/training/api/training_client.py)
-  to manage Training Operator jobs using Python APIs.
+- [Kubeflow Python SDK](https://github.com/kubeflow/training-operator/blob/master/sdk_v2/kubeflow/training/api/training_client.py)
+  to interact with Kubeflow Trainer APIs and to manage TrainJobs.
```
Actually, we want to rename the `kubeflow_training` SDK to `kubeflow` SDK: https://github.com/kubeflow/training-operator/blob/master/sdk_v2/pyproject.toml#L7.
We will not push it to PyPI yet, until we finalize the proposal of creation a new `kubeflow/sdk` repo.
Thus, I prefer we call it Kubeflow SDK in the docs.

varodrig reviewed (commented):
I found Training Operator also in:

- docs/about/community/ under Working groups
- docs/components/katib/overview/
**Diagrams:**
- docs/started/introduction/#kubeflow-overview-diagram
- docs/started/architecture/#kubeflow-ecosystem
- docs/started/architecture/#kubeflow-components-in-the-ml-lifecycle
-

tenzen-y commented on content/en/docs/components/trainer/images/trainer-tech-stack.drawio.svg:
I have same issue with @varodrig

andreyvelich replied:
@varodrig @tenzen-y I updated the diagrams, please can you check again ?

tenzen-y replied:
In my side, the current diagrams work well. Thank you!

andreyvelich commented:
@varodrig 
>docs/about/community/ under Working groups

We keep it Training WG for now, since we don't rename Working Group.

docs/components/katib/overview/
We will update it once we integrate Kubeflow Trainer into Katib

Diagrams:
docs/started/introduction/#kubeflow-overview-diagram
Updated

docs/started/architecture/#kubeflow-ecosystem
Updated

docs/started/architecture/#kubeflow-components-in-the-ml-lifecycle
Updated

andreyvelich commented on content/en/docs/components/trainer/operator-guides/migration.md:
```diff
@@ -0,0 +1,28 @@
+
+## Migration Paths
+
+TODO (andreyvelich): Add docs for migration.
```
We will do that in the followup PR, created this: https://github.com/kubeflow/training-operator/issues/2412

andreyvelich commented on content/en/docs/components/trainer/getting-started.md:
````diff
@@ -0,0 +1,29 @@
+Install the latest Kubeflow Python SDK version directly from the source repository:
+
+```bash
+pip install git+https://github.com/kubeflow/trainer.git@master#subdirectory=sdk_v2
````
Yes, I will update it once we rename the repo.

andreyvelich commented on content/en/_index.html:
```diff
@@ -93,6 +93,25 @@ <h5 class="card-title text-white section-head">Dashboard</h5>
+         <a href="/docs/components/trainer/overview/" target="_blank" rel="noopener" >
+           <img
+             src="/docs/images/logos/kubeflow-trainer.png"
+             class="card-img-top"
+             draggable="false"
+             style="padding: 2rem;"
+             alt="Kubeflow Trainer logo"
+           />
```
@thesuperzapper Do you know what should we adjust to modify padding for the Kubeflow Trainer Icon ?
![Screenshot 2025-02-06 at 14 58 53](https://github.com/user-attachments/assets/1f77b0bc-3e37-4211-8c46-63d5d67b92ae)

andreyvelich commented:
@astefanutti @Electronic-Waste @kubeflow/wg-training-leads @varodrig @saileshd1402 I've updated the Getting Started example.
If that looks good, we should be good to merge this PR.

andreyvelich commented:
We've renamed the repo, and we should be ready to merge this.
/assign @kubeflow/wg-training-leads @Electronic-Waste @varodrig @rimolive @franciscojavierarceo @astefanutti

tenzen-y commented on content/en/docs/components/trainer/images/user-personas.drawio.svg:
I guess that you failed to rebase or squash. The image problem has come back again...

<img width="848" alt="Screenshot 2025-02-12 at 2 10 57" src="https://github.com/user-attachments/assets/415313de-dd30-4246-a6f2-562bf07363a7" />

andreyvelich replied:
Great catch @tenzen-y, please can you check it again ?

tenzen-y reviewed (commented):
Otherwise, lgtm
Thanks for your effort!

andreyvelich reviewed (commented):
Thanks everyone for the feedback 🎉 
/approve
